File: Display.py
import pygame
pygame.init()
COMMANDS = {
    0xAF : ('Display ON (DISON)', ''),
    0xAE : ('Display OFF (DISOFF)', ''),
    0xA6 : ('Normal display (DISNOR)', ''),
    0xA7 : ('Inverse display (DISINV)', ''),
    0xBB : ('Common scan (COMSCAN)', ''),
    0xCA : ('Display control (DISCTL)', ''),
    0x04 : ('Sleep In/Out Preparation (SLPP)', ''),
    0x95 : ('Sleep In (SLPIN)', ''),
    0x94 : ('Sleep Out (SLPOUT)', ''),
    0x75 : ('Page address set (PASET)', ''),
    0x15 : ('Column Address Set (CASET)', ''),
    0xBC : ('Data Scan Direction (DATCTL)', ''),
    0x5C : ('Memory write (RAMWR)', ''),
    0xA8 : ('Partial display in (PLTIN)', ''),
    0xA9 : ('Partial display out (PTLOUT)', ''),
    0xE0 : ('Read modify write in (RMWIN)', ''),
    0xEE : ('Read modify write out (RMWOUT)', ''),
    0xAA : ('Area scroll set (ASCSET)', ''),
    0xAB : ('Scroll start address set (SCSTART)', ''),
    0xD1 : ('Internal OSC on (OSCON)', ''),
    0xD2 : ('Internal OSC off (oSCOFF)', ''),
    0x20 : ('Power Control (PWRCTL)', 'FRAME 1 PWM Set'),
    0x81 : ('Electronic volume control (VOLCTR)', ''),
    0xD6 : ('EC increase 1 (VOLUP)', ''),
    0xD7 : ('EC decrease 1 (VOLDOWN)', ''),
    0x7C : ('Read Register 1 (EPSRRD1)', ''),
    0x7D : ('Read Register 2 (EPSRRD2)', ''),
    0x25 : ('NOP Instruction', ''),
    0x07 : ('EEPROM Function Start (EEOK)', ''),
    0x82 : ('Not Use (RESERVED)', ''),
    0x30 : ('Ext = 0', 'Ext = 0'),
    0x31 : ('Ext = 1', 'Ext = 1'),
    0x23 : ('', 'Palette FRC4 Setup'),
    0x21 : ('', 'Palette FRC2 Setup'),
    0x22 : ('', 'Palette FRC3 Setup'),
    0x32 : ('', 'Analog set (ANASET)'),
    0xCD : ('', 'Control EEPROM'),
    0xCC : ('', 'Cancel EEPROM'),
    0xFC : ('', 'Write data to EEPROM (EPMWR)'),
    0xFD : ('', 'Read data from EEPROM (EPMWR)'),
    0xFA : ('', 'Display performance adjustment (DISPADJ)'),
    0xF4 : ('', 'Internal Initialize Preparation (IIPP)')
    }

class Display():
    def __init__(self):
        self.RedNext = True
        self.RByte = 0
        self.GBByte = 0
        self.pixelPosition = 0
        self.width = 98
        self.height = 67
        self.scale = 2
        
        self.command = 0
        self.drawStartX = 0
        self.drawStartY = 0

        self.maxPixelPosition = self.width*self.height
        self.screen = pygame.display.set_mode((self.width*self.scale, self.height*self.scale))
        pygame.display.set_caption("Miuchiz")
        
    def SendCommand(self, byte):
        self.command = byte
        if byte in COMMANDS:
            command_name = COMMANDS[byte]
        else:
            command_name = 'Unknown'
        print(f'Display command {hex(byte)} ({command_name})')

    def SendData(self, byte):
        if self.command == 0x5C: #Memory Write (RAMWR)
            self.AddPixelData(byte)
        else:
            print(f'Unused data for command {hex(self.command)}: {hex(byte)}')
            
    def AddPixelData(self, byte):
        if self.RedNext:
            self.RByte = byte
        else:
            self.GBByte = byte
            self.DrawCurrentPixel()
       
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                exit()
        self.RedNext = not self.RedNext

    def DrawCurrentPixel(self):
        x = self.pixelPosition % self.width
        y = (self.pixelPosition - x) // self.width
        
        r = self.RByte & 0x0F
        r *= 17
        g = ((self.GBByte & 0xF0) >> 4) * 17
        b = (self.GBByte & 0x0F) * 17

        pygame.draw.rect(self.screen, (r,g,b), (x*self.scale, y*self.scale, self.scale, self.scale))

        self.pixelPosition += 1
        
        if self.pixelPosition >= self.maxPixelPosition:
            self.pixelPosition = 0
            
        pygame.display.flip()

File: Display.cpp
#include "Display.h"

Display::Display(){
    this->RedNext = true;
    this->RByte = 0;
    this->GBByte = 0;
    this->pixelPosition = 0;
    this->width = WIDTH;
    this->height = HEIGHT;
    this->scale = SCALE;
    this->inverse = false;

    this->command = 0;

    this->maxPixelPosition = this->width * this->height;


    SDL_Init(SDL_INIT_EVERYTHING);
    this->window = SDL_CreateWindow("Miuchiz", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, this->width*this->scale, this->height*this->scale, SDL_WINDOW_SHOWN);
    this->renderer = SDL_CreateRenderer(window, -1, 0);
    SDL_SetRenderDrawColor(this->renderer, 0, 255, 0, 255);
    SDL_RenderClear(this->renderer);
}

bool Display::Update(){
    SDL_RenderPresent(this->renderer);
    if (SDL_PollEvent(&this->event) && this->event.type == SDL_QUIT){
        return true;
    }
    return false;
}

void Display::DrawRectangle(BYTE r, BYTE g, BYTE b, unsigned int x, unsigned int y, unsigned int rectWidth, unsigned int rectHeight){
    SDL_SetRenderDrawColor(this->renderer, 0, r, g, b);
    for (unsigned int i=x; i<(x+rectWidth); i++){
        for (unsigned int j=y; j<(y+rectHeight); j++){
            SDL_RenderDrawPoint(this->renderer, i, j);
        }
    }
}

void Display::SendCommand(BYTE by){
    this->command = by;
    if (this->command == 0x75) { //PASET
        this->pixelPosition = 0;
    }
    else if (this->command == 0x15) { //CASET
        this->pixelPosition = 0;
    }
    else if (this->command == 0xA6) { //Normal Display
        this->inverse = false;
    }
    else if (this->command == 0xA7) { //Inverse display
        this->inverse = true;
    }
}

void Display::SendData(BYTE by) {
    if (this->command == 0x5C) { //Memory Write (RAMWR)
        this->AddPixelData(by);
    }
    else {
        printf("Unused data for command %02X: %02X\n", (unsigned int)this->command, (unsigned int)by);
    }
}

void Display::AddPixelData(BYTE by){
    if (this->RedNext) {
        this->RByte = by;
    }
    else {
        this->GBByte = by;
        this->DrawCurrentPixel();
    }
    this->RedNext = !this->RedNext;
}

void Display::DrawCurrentPixel() {
    unsigned int x = this->pixelPosition % this->width;
    unsigned int y = (this->pixelPosition - x) / this->width;

    BYTE r = this->RByte & 0x0F;
    r = r * 17;
    BYTE g = ((this->GBByte & 0xF0) >> 4) * 17;
    BYTE b = (this->GBByte & 0x0F) * 17;

    if (!this->inverse) {
        r = 255 - r;
        g = 255 - g;
        b = 255 - b;
    }

    this->DrawRectangle(r, g, b, x*this->scale, y*this->scale, this->scale, this->scale);

    this->pixelPosition++;

    if (this->pixelPosition >= this->maxPixelPosition) {
        this->pixelPosition = 0;
    }
}

